open-machine-learning-jupyter-book/deep-learning/GAN.md

```diff
@@ -13,5 +13,267 @@ kernelspec:
+
+The aim of Generator is $\bigtriangledown_{W_G} \frac{1}{n} \sum_{i=1}^n log(1-D(G(z^{(i)})))$. 
+$D(G(z^{(i)}))$ is also used to predict fake images, but if it predicts badly on fake images, the probability will be close to 1, which is not same to discriminator.
+
+### Train for convergence
+GAN converges when probabilities are close to 0.5, which means G can cheat D.
+rom a mathematical point of view, it converges when Nash-equilibrium (Game Theory concept) is reached in the minmax (zero-sum) game.
+[Nash-Equilibrium](https://en.wikipedia.org/wiki/Nash_equilibrium) in Game Theory is reached when the actions of one player won't change depending on the opponent's actions.
+
+But not each time GAN can converge, there are some training problems:
+- Oscillation between generator and discriminator loss,
+- Mode collapse (generator produces examples of a particular kind only),
+- Discriminator is too strong, such that the gradient for the generator vanishes and the generator can't keep up,
+- Discriminator is too weak, and the generator produces non-realistic images that fool it too easily (rare problem, though).
+
+For the third problem, replacing $\bigtriangledown_{W_G} \frac{1}{n} \sum_{i=1}^n log(1-D(G(z^{(i)})))$ with $\bigtriangledown_{W_G} \frac{1}{n} \sum_{i=1}^n log(D(G(z^{(i)})))$ can be a good choice.
+
+### Loss
+Discriminator:
+- Maximize prediction probability of classifying real as real and fake as fake,
+- Remember maximizing log likelihood is the same as minimizing negative log likelihood (i.e., minimizing cross-entropy).
+Generator:
+- Minimize likelihood of the discriminator to make correct predictions (predict fake as fake; real as real), which can be achieved by maximizing the cross-entropy,
+- This doesn't work well in practice though because of gradient issues (zero gradient if the discriminator makes correct predictions, which is not what we want for the generator),
+- Better: flip labels and minimize cross entropy (force the discriminator to output high probability for fake if an image is real, and high probability for real if an image is fake).
+
+### Code
+Here we implement a GAN model.
+
+```{code-cell}
+from IPython import display
+
+from torch.utils.data import DataLoader
+from torchvision import transforms, datasets
+
+from utils import Logger
+
+import tensorflow as tf
+import tensorflow.compat.v1 as tf
+tf.disable_v2_behavior() 
+
+import numpy as np
+
+DATA_FOLDER = './tf_data/VGAN/MNIST'
+IMAGE_PIXELS = 28*28
+NOISE_SIZE = 100
+BATCH_SIZE = 100
+
+def noise(n_rows, n_cols):
+    return np.random.normal(size=(n_rows, n_cols))
+
+def xavier_init(size):
+    in_dim = size[0] if len(size) == 1 else size[1]
+    stddev = 1. / np.sqrt(float(in_dim))
+    return tf.random_uniform(shape=size, minval=-stddev, maxval=stddev)
+
+def images_to_vectors(images):
+    return images.reshape(images.shape[0], 784)
+
+def vectors_to_images(vectors):
+    return vectors.reshape(vectors.shape[0], 28, 28, 1)
+```
+
+Load Data

```

Comment:
Is this a title? If it is, make it as a title.

Also, give an empty line before the `{code-cell}`

Reply:
I don't regard it as a title, and I think it just tell us the aim of the following code cell.